### PR TITLE
FIO vb and secure boot for imx7ulpea

### PR DESCRIPTION
**Pull request by ldts**
This PR replaces AVB with FIOVB TA.

### CODE REVIEW

File: drivers/mmc/rpmb.c [hunk 1]
```diff
@@ -223,6 +223,7 @@ static void rpmb_hmac(unsigned char *key, unsigned char *buff, int len,
 	/* finish up 2nd pass */
 	sha256_finish(&ctx, output);
 }
+
```
Comment (ricardosalveti) [inline drivers/mmc/rpmb.c]: No need for this extra line (makes rebasing harder later).

File: drivers/mmc/rpmb.c [hunk 2]
```diff
@@ -380,52 +385,68 @@ static int rpmb_route_write_req(struct mmc *mmc, struct s_rpmb *req,
...
 
-	ret = send_write_mult_block(mmc, req, 1);
+	ret = send_write_mult_block(mmc, req_frame, 1);
 	if (ret)
 		return ret;
 
 	/*
 	 * Read the result of the request.
 	 */
-
```
Comment (ricardosalveti) [inline drivers/mmc/rpmb.c]: Same here.

File: drivers/mmc/rpmb.c [hunk 3]
```diff
@@ -439,28 +460,37 @@ static int rpmb_route_frames(struct mmc *mmc, struct s_rpmb *req,
 	case RPMB_REQ_KEY:
 		if (req_cnt != 1 || rsp_cnt != 1)
 			return -EINVAL;
-		return rpmb_route_write_req(mmc, req, req_cnt, rsp, rsp_cnt);
-
+		ret = rpmb_route_write_req(mmc, req, req_cnt, rsp, rsp_cnt);
+		if (ret)
+			debug("error: RPMB_REQ_KEY\n");
```
Comment (ricardosalveti) [inline drivers/mmc/rpmb.c]: Should this be a debug or an error message? Same for the other checks.

File: cmd/Kconfig [hunk 4]
```diff
@@ -2078,6 +2078,19 @@ config CMD_AVB
...
+config CMD_FIOVB
+	bool "fiovb - FIO Verified Boot 1.0 operations"
+	depends on FIOVB
+	default n
+	help
+	  Enables a "fiovb" command 
+	    fiovb init - initialize avb2 subsystem
+	    fiovb read_rb - read rollback index
+	    fiovb write_rb - write rollback index
+	    fiovb is_unlocked - check device lock state
```
Comment (ricardosalveti) [inline cmd/Kconfig]: As I added to the optee pull request, no need for the lock state here.

File: cmd/Makefile [hunk 5]
```diff
@@ -173,6 +173,9 @@ obj-$(CONFIG_CMD_BLOB) += blob.o
 # Android Verified Boot 2.0
 obj-$(CONFIG_CMD_AVB) += avb.o
 
+# FIO Verified Boot 2.0
```
Comment (ricardosalveti) [inline cmd/Makefile]: 2.0? :-)

Probably good to replace FIO to Foundries.IO here as well.

File: cmd/fiovb.c [hunk 6]
```diff
@@ -0,0 +1,221 @@
+
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: No need for this extra empty line.

File: cmd/fiovb.c [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+/*
+ * (C) Copyright 2018, Linaro Limited
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: Add 2019, Foundries.IO.

File: cmd/fiovb.c [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+ */
+#include <common.h>
+#include <command.h>
+#include <env.h>
+#include <image.h>
+#include <malloc.h>
+#include <mmc.h>
+#include <fiovb.h>
+
+static struct FIOvbOps *fiovb_ops;
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: Is camel case really supported by u-boot?

File: cmd/fiovb.c [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+	mmc_dev = simple_strtoul(argv[1], NULL, 16);
+
+	if (fiovb_ops)
+		fiovb_ops_free(fiovb_ops);
+
+	fiovb_ops = fiovb_ops_alloc(mmc_dev);
+	if (fiovb_ops)
+		return CMD_RET_SUCCESS;
+
+	printf("Failed to initialize fiovb1\n");
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: 1 is the default, just change to 'Failed to initialize fiovb'.

File: cmd/fiovb.c [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	return CMD_RET_FAILURE;
+}
+
+int do_fiovb_read_rb(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
+{
+	size_t index;
+	u64 rb_idx;
+
+	if (!fiovb_ops) {
+		printf("FIOVB 1.0 is not initialized, run 'fiovb init' first\n");
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: Remove 1.0 here, not needed.

File: cmd/fiovb.c [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	return CMD_RET_FAILURE;
+}
+
+int do_fiovb_write_rb(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
+{
+	size_t index;
+	u64 rb_idx;
+
+	if (!fiovb_ops) {
+		printf("FIOVB 1.0 is not initialized, run 'fiovb init' first\n");
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: Same here.

File: cmd/fiovb.c [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+		       char * const argv[])
+{
+	const char *name;
+	size_t bytes;
+	size_t bytes_read;
+	void *buffer;
+	char *endp;
+
+	if (!fiovb_ops) {
+		printf("FIOVB 1.0 is not initialized, run 'avb init' first\n");
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: Here as well. Message is also saying avb instead of fiovb.

File: cmd/fiovb.c [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+int do_fiovb_write_pvalue(cmd_tbl_t *cmdtp, int flag, int argc,
+			char * const argv[])
+{
+	const char *name;
+	const char *value;
+
+	if (!fiovb_ops) {
+		printf("AVB 2.0 is not initialized, run 'avb init' first\n");
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: s/avb/fio/g and drop 2.0.

File: cmd/fiovb.c [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+	if (flag == CMD_FLAG_REPEAT)
+		return CMD_RET_FAILURE;
+
+	return cp->cmd(cmdtp, flag, argc, argv);
+}
+
+U_BOOT_CMD(
+	fiovb, 29, 0, do_fiovb,
+	"Provides commands for testing Android Verified Boot 2.0 functionality",
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: "for testing Foundries.IO verified boot".

File: cmd/fiovb.c [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+	"init <dev> - initialize avb2 for <dev>\n"
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: s/avb2/fiovb/

File: cmd/fiovb.c [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+	"avb read_rb <num> - read rollback index at location <num>\n"
+	"avb write_rb <num> <rb> - write rollback index <rb> to <num>\n"
+	"avb is_unlocked - returns unlock status of the device\n"
+	"avb read_pvalue <name> <bytes> - read a persistent value <name>\n"
+	"avb write_pvalue <name> <value> - write a persistent value <name>\n"
```
Comment (ricardosalveti) [inline cmd/fiovb.c]: Helper all saying 'avb' instead of fiovb.

File: common/fiovb.c [hunk 17]
```diff
@@ -0,0 +1,329 @@
+/*
+ * (C) Copyright 2018, Linaro Limited
```
Comment (ricardosalveti) [inline common/fiovb.c]: Add our own copyright.

File: include/fiovb.h [hunk 18]
```diff
@@ -0,0 +1,153 @@
+
```
Comment (ricardosalveti) [inline include/fiovb.h]: Remove the first empty line.

File: include/fiovb.h [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+/*
+ * (C) Copyright 2018, Linaro Limited
```
Comment (ricardosalveti) [inline include/fiovb.h]: Add our own copyright.

File: include/tee/optee_ta_fiovb.h [hunk 20]
```diff
@@ -0,0 +1,65 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/* Copyright (c) 2018, Linaro Limited */
```
Comment (ricardosalveti) [inline include/tee/optee_ta_fiovb.h:2]: Missing our own copyright.

File: configs/mx7ulpea-ucom_lmp_defconfig [hunk 21]
```diff
@@ -37,10 +38,12 @@ CONFIG_CMD_MEMTEST=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
+# CONFIG_CMD_MMC_RPMB is not set
```
Comment (ricardosalveti) [inline configs/mx7ulpea-ucom_lmp_defconfig]: Do we need to disable this?

File: configs/mx7ulpea-ucom_lmp_defconfig [hunk 22]
```diff
@@ -73,7 +77,9 @@ CONFIG_FSL_LPUART=y
 CONFIG_SPI=y
 CONFIG_DM_SPI=y
 CONFIG_FSL_QSPI=y
+CONFIG_TEE=y
 CONFIG_OPTEE=y
+# CONFIG_OPTEE_TA_AVB is not set
```
Comment (ricardosalveti) [inline configs/mx7ulpea-ucom_lmp_defconfig]: Same here, is this on by default?

File: common/Kconfig [hunk 23]
```diff
@@ -937,7 +937,7 @@ config HASH
 
 config AVB_VERIFY
 	bool "Build Android Verified Boot operations"
-	depends on LIBAVB && FASTBOOT
```
Comment (ricardosalveti) [inline common/Kconfig]: We can drop this patch as we will not be using avb anymore.

File: common/avb_verify.c [hunk 24]
```diff
@@ -356,7 +356,7 @@ static struct mmc_part *get_partition(AvbOps *ops, const char *partition)
 	if (!part)
 		return NULL;
 
-	dev_num = get_boot_device(ops);
+	dev_num = avb_get_boot_device(ops);
```
Comment (ricardosalveti) [inline common/avb_verify.c]: Also another patch we can drop if we are not using.

File: configs/mx7ulpea-ucom_lmp_defconfig [hunk 25]
```diff
@@ -8,6 +8,7 @@ CONFIG_NR_DRAM_BANKS=1
 CONFIG_SPL_STACK_R_ADDR=0x68800000
 CONFIG_ARMV7_BOOT_SEC_DEFAULT=y
 CONFIG_SPL_TEXT_BASE=0x2F011000
+CONFIG_ANDROID_BOOT_IMAGE=y
```
Comment (ricardosalveti) [inline configs/mx7ulpea-ucom_lmp_defconfig]: Also fine to drop as we're not going to use it.

File: cmd/fiovb.c (hunk 7 above)
Comment (ldts) [reply cmd/fiovb.c]: add another copyright line?

File: cmd/fiovb.c (hunk 8 above)
Comment (ldts) [reply cmd/fiovb.c]: nah I dont think so. looks horrible to me but that is what was merged so I thought it was better just to follow this as a guideline. anyway, will change it.

File: configs/mx7ulpea-ucom_lmp_defconfig (hunk 21 above)
Comment (ldts) [reply configs/mx7ulpea-ucom_lmp_defconfig]: no we dont need to. ok

File: configs/mx7ulpea-ucom_lmp_defconfig (hunk 22 above)
Comment (ldts) [reply configs/mx7ulpea-ucom_lmp_defconfig]: seems to be set when I select some other option

File: common/Kconfig (hunk 23 above)
Comment (ldts) [reply common/Kconfig]: sure. I thought we needed avb for the provisioning of the keys (although now that I remember we are only going to be provisioning from linux) ok

File: board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c [hunk 26]
```diff
@@ -95,7 +298,6 @@ int board_init(void)
 #ifdef CONFIG_FSL_QSPI
 	board_qspi_init();
 #endif
-
```
Comment (mike-scott) [inline board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c]: Do we need this line removed?

File: board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c [hunk 27]
```diff
@@ -13,11 +13,25 @@
 #include <asm/arch/iomux.h>
 #include <asm/mach-imx/boot_mode.h>
 #include <asm/gpio.h>
+#include <fs.h>
 #include <usb.h>
 #include <dm.h>
+#include <spi_flash.h>
```
Comment (mike-scott) [inline board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c]: Should we have these protected with? `#ifdef CONFIG_FPGA` -- I know we're enabling the CONFIG in `include/configs/mx7ulpea-ucom.h` but we do an `#ifdef` check below in the source.

File: drivers/tee/optee/rpmb.c [hunk 28]
```diff
@@ -106,6 +108,10 @@ static u32 rpmb_get_dev_info(u16 dev_id, struct rpmb_dev_info *info)
 		return TEE_ERROR_GENERIC;
 
 	memcpy(info->cid, mmc->cid, sizeof(info->cid));
+
+	for (i = 0; i < 4; i++)
```
Comment (mike-scott) [inline drivers/tee/optee/rpmb.c]: I feel like the `4` here should be the array length of info->cid (# of u32 elements)

File: drivers/mmc/rpmb.c [hunk 29]
```diff
@@ -380,52 +384,69 @@ static int rpmb_route_write_req(struct mmc *mmc, struct s_rpmb *req,
...
+	ALLOC_CACHE_ALIGN_BUFFER(struct s_rpmb, rsp_frame, rsp_cnt);
+	ALLOC_CACHE_ALIGN_BUFFER(struct s_rpmb, req_frame, req_cnt);
 	int ret;
 
+	memcpy(req_frame, req, sizeof(struct s_rpmb));
+
 	/*
 	 * Send the read request.
 	 */
+
```
Comment (mike-scott) [inline drivers/mmc/rpmb.c]: Do we need this extra line?

**Review by mike-scott — CHANGES_REQUESTED** [review]
Also:
commit: `[FIO toup] rpmb: use cache aligned buffers on route commands` needs a better description and probably split into 2-3 commits:
- handle aligned buffer access
- there are quite a few MMC cmd request changes like `MMC_RSP_R1` -> `MMC_RSP_R1b` which also need explanation

File: cmd/fiovb.c [hunk 30]
```diff
@@ -0,0 +1,193 @@
...
+
+	if (!fiovb_ops) {
+		printf("Foundries.IO Verified Boot is not initialized, run 'fiovb init' first\n");
+		return CMD_RET_FAILURE;
+	}
+
+	if (argc != 3)
+		return CMD_RET_USAGE;
+
+	index = (size_t)simple_strtoul(argv[1], NULL, 16);
```
Comment (ldts) [inline cmd/fiovb.c]: needs fixing to simple_strtoull

File: cmd/fiovb.c [hunk 31]
```diff
@@ -0,0 +1,193 @@
...
+	rb_idx = simple_strtoul(argv[2], NULL, 16);
```
Comment (ldts) [inline cmd/fiovb.c]: bug. needs strtoull

File: drivers/tee/optee/rpmb.c (hunk 28 above)
Comment (ldts) [reply drivers/tee/optee/rpmb.c]: yep, but array length of mmc->cid instead

Comment (ldts) [reply drivers/tee/optee/rpmb.c]: I'll write it differently and post it upstream

File: board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c (hunk 27 above)
Comment (ldts) [reply board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c]: we can follow that path (@ricardosalveti  did that with the SPL include) but IMO it makes the code a bit uglier and for not much gain...anyhow let me know if you think you'd better see that for consistency

Comment (ricardosalveti) [reply board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c]: Yeah, seems the standard in u-boot is to just include the header when it is building for SPL, so probably better to protect here as well.

File: board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c (hunk 26 above)
Comment (ricardosalveti) [reply board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c]: Had the same comment, @ldts mind removing the removal?

File: configs/mx7ulpea-ucom_lmp_defconfig [hunk 32]
```diff
@@ -21,6 +21,7 @@ CONFIG_BOOTCOMMAND="fatload mmc 0:1 ${loadaddr} /boot.itb; setenv verify 1; sour
 CONFIG_DEFAULT_FDT_FILE="imx7ulpea-ucom-kit_v2"
 CONFIG_BOUNCE_BUFFER=y
 CONFIG_BOARD_EARLY_INIT_F=y
+CONFIG_FIOVB=y
```
Comment (ricardosalveti) [inline configs/mx7ulpea-ucom_lmp_defconfig]: Please also add the same changes to configs/mx7ulpea-ucom-mfgtool_defconfig, as that is the one used by the mfgtool build.

**Review by mike-scott — APPROVED** [review]
LGTM

**Pull request merged**